### PR TITLE
METRON-1820 Syslog support for new api -> multiple messages and errors

### PR DESCRIPTION
**Pull request by ottobackwards**
This PR adds support for the Syslog parser and having multiple messages and errors.

Run a syslog 5424 source, with some invalid messages mixed in, the valid messages should still get indexed.

In order to streamline the review of the contribution we ask you follow these guidelines and ask you to double check the following:

### For all changes:
- [x] Is there a JIRA ticket associated with this PR? If not one needs to be created at [Metron Jira](https://issues.apache.org/jira/browse/METRON/?selectedTab=com.atlassian.jira.jira-projects-plugin:summary-panel).
- [x] Does your PR title start with METRON-XXXX where XXXX is the JIRA number you are trying to resolve? Pay particular attention to the hyphen "-" character.
- [x] Has your PR been rebased against the latest commit within the target branch (typically master)?

### For code changes:
- [ ] Have you included steps to reproduce the behavior or problem that is being changed or addressed?
- [x] Have you included steps or a guide to how the change may be verified and tested manually?
- [x] Have you ensured that the full suite of tests and checks have been executed in the root metron folder via:
  ```
  mvn -q clean integration-test install && dev-utilities/build-utils/verify_licenses.sh 
  ```

- [x] Have you written or updated unit tests and or integration tests to verify your changes?
- [x] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)?
- [ ] Have you verified the basic functionality of the build by building and running locally with Vagrant full-dev environment or the equivalent?

### For documentation related changes:
- [ ] Have you ensured that format looks appropriate for the output in which it is rendered by building and verifying the site-book? If not then run the following commands and the verify changes via `site-book/target/site/index.html`:

  ```
  cd site-book
  mvn site
  ```

#### Note:
Please ensure that once the PR is submitted, you check travis-ci for build issues and submit an update to your PR as soon as possible.
It is also recommended that [travis-ci](https://travis-ci.org) is set up for your personal repository such that your branches are built there before submitting a pull request.

### CODE REVIEW

**Comment by JonZeolla** [comment]
+1 via inspection of this and palindromicity/simple-syslog-5424#18, pending travis

**Comment by ottobackwards** [comment]
I had missed support for the master throwable in the old List<JSONObject> parse function.  This function is not called by the Bolt anymore, but I corrected the issue in this and Grok.

File: metron-platform/metron-parsers/src/main/java/org/apache/metron/parsers/syslog/Syslog5424Parser.java [hunk 1]
```diff
@@ -61,16 +67,37 @@ public void configure(Map<String, Object> config) {
   public void init() {
   }
 
-  @Override
   @SuppressWarnings("unchecked")
+  @Override
   public List<JSONObject> parse(byte[] rawMessage) {
+    Optional<MessageParserResult<JSONObject>> resultOptional = parseOptionalResult(rawMessage);
```
Comment (mmiklavc) [inline metron-platform/metron-parsers/src/main/java/org/apache/metron/parsers/syslog/Syslog5424Parser.java]: Based on the changes to the ParserBolt from https://github.com/apache/metron/pull/1184, where would this still be used?

Comment (ottobackwards) [reply metron-platform/metron-parsers/src/main/java/org/apache/metron/parsers/syslog/Syslog5424Parser.java]: It wouldn't, but the interface must be implemented.  It is awkward, I'm not sure how we want to proceed with that overall.  When we added the first parseOptional and moved the bolt to it we should have deprecated the old method or something....

Comment (ottobackwards) [reply metron-platform/metron-parsers/src/main/java/org/apache/metron/parsers/syslog/Syslog5424Parser.java]: I'm open to a different approach, I could throw a notimplementedexception or something.

Comment (mmiklavc) [reply metron-platform/metron-parsers/src/main/java/org/apache/metron/parsers/syslog/Syslog5424Parser.java]: Just spitballing here, but maybe an option here is to create a decorator interface around the older implementation and provide defaults when someone implements the new interface? Old api doesn't change and we handle new implementations gracefully.

```java
public interface MessageParser<T> extends Configurable {
  List<T> parse(byte[] rawMessage);
  Optional<List<T>> parseOptional(byte[] parseMessage);
}
public interface MultilineMessageParser<T> extends MessageParser {
  default List<T> parse(byte[] rawMessage) {
    throw new NotImplementedException();
  }
  default Optional<MessageParserResult<T>> parseOptionalResult(byte[] parseMessage) {
    List<T> list = new ArrayList<>();
      try {
        Optional<List<T>> optionalMessages = parseOptional(parseMessage);
        optionalMessages.ifPresent(list::addAll);
      } catch (Throwable t) {
        return Optional.of(new DefaultMessageParserResult<>(t));
      }
      return Optional.of(new DefaultMessageParserResult<T>(list));
  }
}
```

**ParserBolt:**
```java
MultilineMessageParser mmp = null;
if (!(parser instanceof MultilineMessageParser)) {
  mmp = new MultilineMessageParser<JSONObject>() {

    @Override
    public void configure(Map<String, Object> config) {
      parser.configure(config);
    }

    @Override
    public void init() {
      parser.init();
    }

    @Override
    public boolean validate(JSONObject message) {
      return parser.validate(message);
    }
  };
} else {
  mmp = (MultilineMessageParser) parser;
}

Optional<MessageParserResult<JSONObject>> results = mmp.parseOptionalResult(rawMessage.getMessage());
```

Comment (ottobackwards) [reply metron-platform/metron-parsers/src/main/java/org/apache/metron/parsers/syslog/Syslog5424Parser.java]: I think that is a fine idea, I had thought of a new interface, but I decided to follow the precedent of parseOptional.

**Comment by mmiklavc** [comment]
lgtm @ottobackwards. +1 by inspection.

**Pull request closed without merging**